airflow/jobs/scheduler_job_runner.py

```diff
@@ -1838,12 +1890,63 @@ def _orphan_unreferenced_datasets(self, session: Session = NEW_SESSION) -> None:
         _executor_to_tis: defaultdict[BaseExecutor, list[TaskInstance]] = defaultdict(list)
-        executor: str | None
         for ti in tis:
-            if ti.executor:
-                executor = str(ti.executor)
+            if executor_obj := self._try_to_load_executor(ti.executor):
+                _executor_to_tis[executor_obj].append(ti)
             else:
-                executor = None
-            _executor_to_tis[ExecutorLoader.load_executor(executor)].append(ti)
+                continue
 
         return _executor_to_tis
+
+    def _slots_free_for_tis(self, tis: list[TaskInstance]) -> set[TaskInstance]:
+        """Return TIs that we have slots available for."""
+        # First get a mapping of executor names to slots they have available
+        executor_to_slots_available: dict[ExecutorName, int] = {}
+        for executor in self.job.executors:
+            if TYPE_CHECKING:
+                # All executors should have a name if they are initted from the executor_loader. But we need
+                # to check for None to make mypy happy.
+                assert executor.name
+            executor_to_slots_available[executor.name] = executor.slots_available
+
+        # Loop through all the TIs we're scheduling for and add them to a set to be moved to queued if the
+        # executor they're assigned to has slots available.
+        tis_we_have_room_for = set()
+        for ti in tis:
+            if executor_obj := self._try_to_load_executor(ti.executor):
+                ti_exec_name = executor_obj.name
+                if TYPE_CHECKING:
+                    # All executors should have a name if they are initted from the executor_loader. But we
+                    # need to check for None to make mypy happy.
+                    assert ti_exec_name
+                if executor_to_slots_available[ti_exec_name] > 0:
+                    tis_we_have_room_for.add(ti)
+                    executor_to_slots_available[ti_exec_name] -= 1
+            else:
+                continue
+
+        return tis_we_have_room_for
+
+    def _try_to_load_executor(self, executor_name: str | Column | None) -> BaseExecutor | None:
+        """Try to load the given executor.
+
+        In this context, we don't want to fail if the executor does not exist. Catch the exception and
+        log to the user.
+        """
+        # Executor name comes in as a column type which upsets mypy. Here we are "casting" the value to
+        # either None or a string to satisfy typing
+        if not executor_name:
+            executor_name = None
+        else:
+            executor_name = str(executor_name)
+        try:
+            return ExecutorLoader.load_executor(executor_name)
+        except AirflowException as e:
+            if "Unknown executor" in str(e):
+                self.log.warning(
+                    "Executor, %s, was not found but a Task was configured to use it", executor_name
+                )
```

Comment:
this isn't the most helpful message because we don't know what task was configured to use it... is there somewhere that a user would be able to see that they set a bad executor on their task?  

one thing you should probably do is use task context logger to send a message to the TI logs.  but ....  what even happens to a task that is misconfigured in this way... will the scheduler keep retrying it continuously?  then we might get _too many_ such messages sent to the TI logs.  this is sorta why it feels like we should probably fail the task in that scenario.  if the executor is not enabled, what chance is there that it will be scheduled?  why should the scheduler keep retrying that task?

Reply:
As I said in the other thread (maybe we should got back to that one so the context is all in one spot?) this should not really happen. The dag parser will not be able to validate the dag, and it will tell you what task has an executor that is misconfigured and you'll get a banner in the UI (like when your dag has a python error), so the task will not get scheduled. I thought I'd put a guard here just in case, since it's pretty critical if you do somehow arrive here and the user should get (at least some) feedback about that.

Reply:
I can add a comment above this log explaining that context. I think that's worth doing so that it doesn't confuse others in the future :+1:

Reply:
sorry if i missed your earlier explanation.  so you are saying it the dag will fail to load if it has a task that specifies an executor that is not supported by configuration.  roger 👍